src/core/IO/Socket/Async.pm6

```diff
@@ -165,42 +166,65 @@ my class IO::Socket::Async {
-        my $p = Promise.new;
+        my Promise $p .= new;
         my $v = $p.vow;
-        my $encoding = Encoding::Registry.find($enc);
+        my \server-socket = await create-socket(:$scheduler);
         nqp::asyncconnect(
             $scheduler.queue,
-            -> Mu \socket, Mu \err, Mu \peer-host, Mu \peer-port, Mu \socket-host, Mu \socket-port {
+            -> Mu \client-socket, Mu \err, Mu \peer-host, Mu \peer-port, Mu \socket-host, Mu \socket-port {
                 if err {
-                    $v.break(err);
-                }
-                else {
-                    my $client_socket := nqp::create(self);
-                    nqp::bindattr($client_socket, IO::Socket::Async, '$!VMIO', socket);
-                    nqp::bindattr($client_socket, IO::Socket::Async, '$!enc', $encoding.name);
-                    nqp::bindattr($client_socket, IO::Socket::Async, '$!encoder',
-                        $encoding.encoder());
-                    nqp::bindattr($client_socket, IO::Socket::Async, '$!peer-host', peer-host);
-                    nqp::bindattr($client_socket, IO::Socket::Async, '$!peer-port', peer-port);
-                    nqp::bindattr($client_socket, IO::Socket::Async, '$!socket-host', socket-host);
-                    nqp::bindattr($client_socket, IO::Socket::Async, '$!socket-port', socket-port);
-
-                    setup-close($client_socket);
-                    $v.keep($client_socket);
+                    $v.break: err;
+                } else {
+                    my \client   = nqp::create(self);
+                    my $encoding = Encoding::Registry.find($enc);
+                    nqp::bindattr(client, IO::Socket::Async, '$!VMIO', client-socket);
+                    nqp::bindattr(client, IO::Socket::Async, '$!enc', $encoding.name);
+                    nqp::bindattr(client, IO::Socket::Async, '$!encoder', $encoding.encoder());
+                    nqp::bindattr(client, IO::Socket::Async, '$!peer-host', peer-host);
+                    nqp::bindattr(client, IO::Socket::Async, '$!peer-port', peer-port);
+                    nqp::bindattr(client, IO::Socket::Async, '$!socket-host', socket-host);
+                    nqp::bindattr(client, IO::Socket::Async, '$!socket-port', socket-port);
+                    setup-close(client);
+                    $v.keep: client;
                 }
             },
-            $host, $port, SocketCancellation);
+            server-socket, $host, $port, SocketCancellation);
         $p
     }
 
     class ListenSocket is Tap {
+        has         $!VMIO;
         has Promise $.socket-host;
         has Promise $.socket-port;
 
-        method new(&on-close, Promise :$socket-host, Promise :$socket-port) {
-            self.bless(:&on-close, :$socket-host, :$socket-port);
+        submethod BUILD(Mu :$!VMIO, Promise :$!socket-host, Promise :$!socket-port) { }
+
+        method new(&on-close, Mu :$VMIO, Promise :$socket-host, Promise :$socket-port) {
```

Comment:
Should `$VMIO be required` ala `Mu :$VMIO!`?

Reply:
I don't see why not